p4include/v1model.p4

```diff
@@ -250,8 +280,30 @@ extern action_profile {
  */
 extern void random<T>(out T result, in T lo, in T hi);
 
-// If the type T is a named struct, the name is used
-// to generate the control-plane API.
+/***
+ * Calling digest causes a message containing the values specified in
+ * the data parameter to be sent to the control plane software.  It is
+ * similar to sending a clone of the packet to the control plane
+ * software, except that it can be more efficient because the messages
+ * are typically smaller than packets, and many such small digest
+ * messages are typically coalesced together into a larger "batch"
+ * which the control plane software processes all at once.
+ *
+ * Calling digest is only supported in the ingress control.  There is
+ * no way to undo its effects once it has been called.
+ *
+ * If the type T is a named struct, the name is used to generate the
+ * control plane API.
+ *
+ * The BMv2 implementation of the v1model architecture ignores the
+ * value of the receiver parameter.
+ *
+ * TBD As of 2019-Apr, the current BMv2 simple_switch implementation
```

Comment:
If this is true then we have to update the resubmit patch to also handle digest in a similar way.

Reply:
I am pretty sure this is true, but can double check with @cc10512 @hanw on this detail to see if it is correct for P4_14.

Reply:
for the bmv2 behavior @antoninbas is likely a better resource!

Reply:
@mbudiu-vmw I have updated this PR to say that it is definitely the case that the values sent to the control plane because of the P4 program executing a digest call are the values those fields have at the end of execution of the ingress control, _not_ at the time the digest call is made, so it is like recirculate, resubmit, and clone3 operations in that regard, although it does not cause any values to be "restored" at the beginning of ingress or egress processing the way those other operations do.

@hanw confirmed this in private communication, and I have also examined the BMv2 simple_switch implementation to verify that it does this for P4_14 programs.

Reply:
Actually I believe that the compiler will do the right thing and copy these values to a temporary. So the semantics you get is that the fields are digested right away. This means that the comment is incorrect.

Reply:
The comment is correct for P4_14 behavior, which v1model is supposed to implement, I thought.

If P4_14 behavior of digest creation is to wait until the end of ingress before packaging up the specified field_list fields into a message to the control plane software, then how is the comment incorrect?

Reply:
I don't think that the P4-14 spec mandates a specific behavior.

Reply:
I believe you are right.  I should have said this instead:

+ @hanw said that he knows of a prominent implementation that captures digest field values at the end of the ingress pipeline, and
+ BMv2 simple_switch has also done it this way for years when compiling P4_14 programs.

Thus if P4_16 + v1model captures the field values at the time of the digest() call, it could change the behavior of auto-translated P4_14 programs, if the auto-translation does the straighforward thing of replacing a digest call in P4_14 with a corresponding digest call at the same place in the code execution flow of the translated P4_16 program.

Reply:
@mbudiu-vmw Let me know if you believe anything about how the digest operation is documented in the latest PR looks wrong to you.  As far as I can tell, based on my previous comment, it is correct.

Reply:
The p4-14 to p4-16 translator has always used the alternative semantics for digest, where the operation happens right away. This happens to work for p4 programs that do not modify the digested metadata after the digest call. If we had this specification earlier we could have implemented the other semantics - but doing this now may break user programs that depend on the current implementation.

So which programs are "right" - the p4-14 programs or the existing p4-16 ones?

Reply:
I don't know.  Just wanted to make sure people were aware of differences in behavior here.

I could change the documentation for digest to mention the difference between P4_14 and P4_16+v1model behavior, and warn people that the safest thing is to never change values given to a digest call after the digest call is made, if they want to avoid this behavior difference.

Reply:
@mbudiu-vmw I have updated the documentation about the behavior of digest, and how it can differ between the current P4_16+v1model implementation, vs. some P4_14 implementations.  Let me know if it looks reasonable to you.

Reply:
@mbudiu-vmw I think this is ready to merge in now.  Feel free to take a look at the current latest contents of the documentation on the digest function, which explicitly mentions how it behaves (create digest message contents from the current values of the data parameter), and also that some other implementations of P4_14 and, and perhaps some other P4_16+v1model implementations, might instead use the values of that parameter at the end of the ingress pipeline.

Reply:
I would have wanted to merge first my PR about recirculate, since otherwise I will have to rebase it again, and that's very painful. I am hoping someone will approve it.

Reply:
Understood.  I had forgotten that PR also makes changes to v1model.p4.  I'm happy for this one to wait a while longer, since it is only documentation comment changes.

For that other one to make progress, it seems like a few more people will need to express some kind of opinion on what approaches are viable, and which are not, and why.